src/notifications/model.rs

```diff
@@ -183,13 +183,194 @@ impl Notifications {
     }
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SetNotificationStatePayload {
+    ids: Vec<i32>,
+    user_id: i32,
+    unread: bool,
+}
+
+#[derive(Error, Debug)]
+pub enum SetNotificationStateError {
+    #[error("Notification state cannot be set because of a database error: {inner:?}.")]
+    DatabaseError { inner: sqlx::Error },
+    #[error("Notifications cannot be fetched because of a database error: {inner:?}.")]
+    NotificationsError { inner: NotificationsError },
+}
+
+impl From<sqlx::Error> for SetNotificationStateError {
+    fn from(inner: sqlx::Error) -> Self {
+        SetNotificationStateError::DatabaseError { inner }
+    }
+}
+
+impl From<NotificationsError> for SetNotificationStateError {
+    fn from(error: NotificationsError) -> Self {
+        match error {
+            NotificationsError::DatabaseError { inner } => inner.into(),
+        }
+    }
+}
+
+impl Notifications {
+    pub async fn set_notification_state<'a, E>(
+        payload: SetNotificationStatePayload,
+        executor: E,
+        pool: &MySqlPool,
+    ) -> Result<Notifications, SetNotificationStateError>
+    where
+        E: Executor<'a>,
+    {
+        let mut transaction = executor.begin().await?;
+
+        for id in payload.ids.into_iter() {
+            let seen = !payload.unread;
+            let notification = sqlx::query!(
+                r#"
+                    SELECT id, seen
+                        FROM notification
+                        WHERE id = ?
+                "#,
+                id
+            )
+            .fetch_one(&mut transaction)
+            .await?;
+
+            if (notification.seen != 0) == seen {
+                continue;
+            }
+
+            sqlx::query!(
+                r#"
+                    UPDATE notification
+                        SET seen = ?
+                        WHERE id = ?
+                "#,
+                seen,
+                id
+            )
+            .execute(&mut transaction)
+            .await?;
```

Comment:
We probably don't need the `SELECT` beforehand. I initially did that because I wanted to track which rows did change but didn't need it later. We might even be able to do this in a single statement.